### PR TITLE
[jenkins] fix: compiler and OS update for catapult client

### PR DESCRIPTION
**Pull request by Wayonb**
## What's the issue?
Catapult builds are using older version of compilers.

## How have you changed the behavior?
Latest gcc and clang compiler are used for catapult client builds.

## How was this change tested?
Tested with Jenkins.

There are still some failures I am looking into with gcc 12 and sans tests

### CODE REVIEW

File: jenkins/catapult/compilers/ubuntu-clang/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Use the latest Ubuntu 22.04 image.
```
Comment (Jaguar0625) [inline jenkins/catapult/compilers/ubuntu-clang/Dockerfile:1]: good job parameterizing 👍

File: jenkins/catapult/configurations/gcc-code-coverage.yaml [hunk 2]
```diff
@@ -1,3 +1,3 @@
-compiler: ../compilers/gcc-10.yaml
+compiler: ../compilers/gcc-11.yaml
```
Comment (Jaguar0625) [inline jenkins/catapult/configurations/gcc-code-coverage.yaml:1]: 1. should this be gcc latest (12) ?
2. should we have a clang-latest.yaml gcc-latest.yaml to minimize future changes when bumping versions ?

File: jenkins/catapult/configurations/gcc-westmere.yaml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+compiler: ../compilers/gcc-11.yaml
```
Comment (Jaguar0625) [inline jenkins/catapult/configurations/gcc-westmere.yaml:1]: should this be gcc latest (12)?

File: jenkins/catapult/jenkins/catapult-client-build-compiler-image.groovy [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+						echo """
+									env.GIT_BRANCH: ${env.GIT_BRANCH}
+								 MANUAL_GIT_BRANCH: ${MANUAL_GIT_BRANCH}
+
+							COMPILER_VERSION: ${COMPILER_VERSION}
+								  OPERATING_SYSTEM: ${OPERATING_SYSTEM}
+
+								   destImageName: ${destImageName}
+						"""
+					}
```
Comment (Jaguar0625) [inline jenkins/catapult/jenkins/catapult-client-build-compiler-image.groovy:47]: colon align printed lines?

File: jenkins/catapult/templates/UbuntuTestSanitizerBaseImage.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,21 @@
+FROM symbolplatform/symbol-server-compiler:ubuntu-clang-14
+ARG DEBIAN_FRONTEND=noninteractive
```
Comment (Jaguar0625) [inline jenkins/catapult/templates/UbuntuTestSanitizerBaseImage.Dockerfile:2]: 1. ARG or ENV ?
2. can't you use UbuntuTestBaseImage.dockerfile?

File: jenkins/catapult/baseImageDockerfileGenerator.py [hunk 6]
```diff
@@ -131,6 +131,9 @@ def rocks(self):
 		descriptor.options += DEPENDENCY_FLAGS['facebook_rocksdb']
 		descriptor.options += ['-DUSE_RTTI=1']
 
+		if self.compiler.c.startswith('gcc') and self.compiler.version == 12:
+			descriptor.options += ['-DFAIL_ON_WARNINGS=OFF']
```
Comment (Jaguar0625) [inline jenkins/catapult/baseImageDockerfileGenerator.py:135]: 1. add comment about why we're (temporarily) disabling
2. can you just disable specific warnings (instead of all)

File: jenkins/catapult/baseImageDockerfileGenerator.py [hunk 7]
```diff
@@ -355,8 +358,26 @@ def add_git_dependency(organization, project, versions_map, options, revision=1)
...
+		'make install',
+		'cd ..',
+		'rm -rf openssl'
+	], OPEN_SSL_OPTIONS=' '.join(options.openssl()), OPEN_SSL_CONFIGURE=' '.join(configure), VERSION=version, COMPILER=compiler)
+
+
 def generate_phase_deps(options):
 	print(f'FROM {options.layer_image_name("boost")}')
+
+	add_openssl(options, [])
```
Comment (Jaguar0625) [inline jenkins/catapult/baseImageDockerfileGenerator.py:379]: why are you always installing openssl now ?

File: jenkins/catapult/runDockerBuildInnerPrepare.py [hunk 8]
```diff
@@ -31,12 +31,6 @@ def main():
 	for name in bin_folder_names:
 		environment_manager.copy_tree_with_symlinks(DATA_VOLUME / 'binaries' / name, USER_HOME / name)
 
-	# LD_LIBRARY_PATH is not passed when llvm-symbolizer is started via asan, so move libs to system location
-	if is_dev_build:
-		system_bin_path = environment_manager.system_bin_path
-		environment_manager.move_glob_with_symlinks(USER_HOME / 'deps', 'libLLVM*', system_bin_path)
-		process_manager.dispatch_subprocess(['ls', '-laF', system_bin_path])
-
```
Comment (Jaguar0625) [inline jenkins/catapult/runDockerBuildInnerPrepare.py:39]: how was this fixed? i.e. why can it be removed?

File: jenkins/catapult/configurations/gcc-code-coverage.yaml (hunk 2 above)
Comment (Wayonb) [reply jenkins/catapult/configurations/gcc-code-coverage.yaml:1]: 1. Yes, can move it to 12.  All the daily tests should run with the latest compiler?  Weekly tests are the current - 1 compiler version?
2. Good idea, I think that would help

File: jenkins/catapult/templates/UbuntuTestSanitizerBaseImage.Dockerfile (hunk 5 above)
Comment (Wayonb) [reply jenkins/catapult/templates/UbuntuTestSanitizerBaseImage.Dockerfile:2]: 1. ARG is better to use since the variable is only required during the docker build.  ENV will be set when the container is running.
2. Yea, its same except I remove some tools.

File: jenkins/catapult/baseImageDockerfileGenerator.py (hunk 6 above)
Comment (Wayonb) [reply jenkins/catapult/baseImageDockerfileGenerator.py:135]: 1. ok
2. not sure it thats possible from cmake settings.

File: jenkins/catapult/baseImageDockerfileGenerator.py (hunk 7 above)
Comment (Wayonb) [reply jenkins/catapult/baseImageDockerfileGenerator.py:379]: Ubuntu got upgraded from 20.04 -> 22.04.  22.04 ships with openssl 3.x and catapult is still using openssl 1.1.1 so need to install in the base image

File: jenkins/catapult/runDockerBuildInnerPrepare.py (hunk 8 above)
Comment (Wayonb) [reply jenkins/catapult/runDockerBuildInnerPrepare.py:39]: Sanitizer test base image has clang the compiler already installed.  No need to copy compiler files.
https://github.com/symbol/symbol/pull/354/files/3f5d57935b8bb5978627131d4f6b9022404a6f60#diff-579119f97023a7b6e06ea31aaac2573264eac7fd0b2cfe21051fdd638fbbe14a

File: jenkins/catapult/templates/UbuntuTestSanitizerBaseImage.Dockerfile (hunk 5 above)
Comment (Jaguar0625) [reply jenkins/catapult/templates/UbuntuTestSanitizerBaseImage.Dockerfile:2]: `ARG is better` you should check other usages of ARG vs ENV in dockerfiles as a post check-in task

File: jenkins/catapult/configurations/gcc-code-coverage.yaml (hunk 2 above)
Comment (Jaguar0625) [reply jenkins/catapult/configurations/gcc-code-coverage.yaml:1]: >  the daily tests should run with the latest compiler? Weekly tests are the current - 1 compiler version

i think that is fine, @gimre-xymcity

File: client/catapult/tests/catapult/utils/CatapultExceptionTests.cpp [hunk 9]
```diff
@@ -97,7 +98,15 @@ namespace catapult {
 			if (std::string::npos != foundTraits) {
 				auto traitsEnd = functionFullName.find("]", foundTraits);
 				auto functionTraits = functionFullName.substr(foundTraits + 2, traitsEnd - foundTraits - 2);
-				oss << "<" << functionTraits << ">";
+				oss << "<";
+
+	// A bug in gcc 12 where the trait information is missing the namespace.
+	// Gcc 12 returns this - [with TTraits = {anonymous}::RuntimeErrorTraits]
+	#if (__GNUC__ == 12)
```
Comment (Jaguar0625) [inline client/catapult/tests/catapult/utils/CatapultExceptionTests.cpp]: if this is GCC 12 bug, shouldn't it be `#if 12 != __GNUC_`

File: jenkins/catapult/jenkins/catapult-client-build-base-image-all.groovy [hunk 10]
```diff
@@ -43,32 +43,32 @@ pipeline {
...
 						}
 					}
 				}
 
-				stage('clang 13') {
+				stage('clang prior') {
 					steps {
 						script {
-							dispatch_build_base_image_job('clang-13', 'ubuntu', false)
+							dispatch_build_base_image_job('clang-prior', 'ubuntu', true)
```
Comment (Jaguar0625) [inline jenkins/catapult/jenkins/catapult-client-build-base-image-all.groovy:64]: did you intentionally change false to true?

File: jenkins/catapult/jenkins/catapult-client-build-catapult-project.groovy [hunk 11]
```diff
@@ -6,7 +6,8 @@ pipeline {
 	parameters {
 		gitParameter branchFilter: 'origin/(.*)', defaultValue: "${env.GIT_BRANCH}", name: 'MANUAL_GIT_BRANCH', type: 'PT_BRANCH'
 		choice name: 'COMPILER_CONFIGURATION',
-			choices: ['gcc-12', 'gcc-11', 'gcc-westmere', 'clang-13', 'clang-14', 'clang-ausan', 'clang-tsan', 'gcc-code-coverage'],
+			choices: ['gcc-latest', 'gcc-prior', 'gcc-westmere', 'clang-latest', 'clang-prior', 'clang-ausan', 'clang-tsan',
+					  'gcc-code-coverage'],
```
Comment (Jaguar0625) [inline jenkins/catapult/jenkins/catapult-client-build-catapult-project.groovy:10]: are leading spaces intentional ?

File: jenkins/catapult/environment.py [hunk 12]
```diff
@@ -28,6 +28,18 @@ def system_bin_path(self):
 
 		raise RuntimeError('unable to detect system bin path')
 
+	@property
+	def local_lib_path(self):
+		if self.dry_run:
+			return '<LOCAL_LIB_PATH>'
+
+		for descriptor in [('fedora', '/usr/local/lib64'), ('ubuntu', '/usr/local/lib')]:
```
Comment (Jaguar0625) [inline jenkins/catapult/environment.py:36]: i guess we don't call this for debian ?

File: jenkins/catapult/environment.py [hunk 13]
```diff
@@ -28,6 +28,18 @@ def system_bin_path(self):
 
 		raise RuntimeError('unable to detect system bin path')
 
+	@property
+	def local_lib_path(self):
```
Comment (Jaguar0625) [inline jenkins/catapult/environment.py:32]: why did we need to add this ?

File: jenkins/docker/cpp.Dockerfile [hunk 14]
```diff
@@ -1,4 +1,4 @@
-FROM symbolplatform/symbol-server-build-base:ubuntu-gcc-11-conan
+FROM symbolplatform/symbol-server-build-base:ubuntu-gcc-12-conan
```
Comment (Jaguar0625) [inline jenkins/docker/cpp.Dockerfile:1]: should we rename build-base to drop numbered versions too ? 
(can be separate PR ofc)

File: client/catapult/tests/catapult/utils/CatapultExceptionTests.cpp (hunk 9 above)
Comment (Wayonb) [reply client/catapult/tests/catapult/utils/CatapultExceptionTests.cpp]: The issue is that the namespace information is in the exception message but missing from the boost function.
if GCC 12 then add the missing namespace to the trait so it will match the exception info

File: jenkins/catapult/jenkins/catapult-client-build-base-image-all.groovy (hunk 10 above)
Comment (Wayonb) [reply jenkins/catapult/jenkins/catapult-client-build-base-image-all.groovy:64]: yea, I added a conan build for previous compiler version.

File: jenkins/catapult/jenkins/catapult-client-build-catapult-project.groovy (hunk 11 above)
Comment (Wayonb) [reply jenkins/catapult/jenkins/catapult-client-build-catapult-project.groovy:10]: no, editor auto warps :(

File: jenkins/catapult/environment.py (hunk 12 above)
Comment (Wayonb) [reply jenkins/catapult/environment.py:36]: yea, it does get call for Debian, but it will match ubuntu since it checks for path existing.  Not sure if there is a better way.

File: jenkins/catapult/environment.py (hunk 13 above)
Comment (Wayonb) [reply jenkins/catapult/environment.py:32]: Due to openssl getting copied to all the test base images.  Fedora also now ships with openssl 3.
Local lib path is different for Fedora vs Ubuntu.

File: jenkins/docker/cpp.Dockerfile (hunk 14 above)
Comment (Wayonb) [reply jenkins/docker/cpp.Dockerfile:1]: Yea, I was thinking dropping the numbered version also.  The main issue would be that if the latest image is broken during an upgrade then CI will also be broken

File: client/catapult/tests/catapult/utils/CatapultExceptionTests.cpp (hunk 9 above)
Comment (Jaguar0625) [reply client/catapult/tests/catapult/utils/CatapultExceptionTests.cpp]: so, you mean it is a bug in boost?

File: jenkins/catapult/environment.py (hunk 12 above)
Comment (Jaguar0625) [reply jenkins/catapult/environment.py:36]: ok, that's probably fine

File: jenkins/docker/cpp.Dockerfile (hunk 14 above)
Comment (Jaguar0625) [reply jenkins/docker/cpp.Dockerfile:1]: how is that different than if base image is broken now?

File: client/catapult/tests/catapult/utils/CatapultExceptionTests.cpp (hunk 9 above)
Comment (Wayonb) [reply client/catapult/tests/catapult/utils/CatapultExceptionTests.cpp]: maybe. idk.  I thought compiler since it works in gcc11.

File: jenkins/docker/cpp.Dockerfile (hunk 14 above)
Comment (Wayonb) [reply jenkins/docker/cpp.Dockerfile:1]: Currently CI will use a specific base image plus conan packages.  If new compiler(new base image) or packages is been tested on a branch it will not affect the other branches.
If the versioned number is removed then when compiler gets updated, all jobs will be using the same base image.
Its only compiler updates so not a big deal.   I can remove update it in another pr.

**Review by Jaguar0625 — APPROVED** [review]
👍
